mutant schemata

peatersson opened the pull request:
*** Work in progress - Do not merge ***

* Added lib/mutantschemata to Dextool. Two separate parts of mutantschemata
1. source/mutantschemata - D-side of the api. Is intended to work as an api to cpp code for inserting, selecting, deleting etc in the mutation database by utilizing miniorm.
2. cpp_source - Cpp-side of the api along with the implementation of schemata (currently placed in rewrite.hpp). "Links" external functions in the api together with the D-side of the code along with utility functionality.

* Changed compileflags for Dextool (impact of this needs to be further investigated)
-cppflags -> -cxxflags

* Has modified plugin/mutate/frontend/frontend.d to call mutantschemata in the analyze phase. This needs to be removed in order to not impact original functionality of Dextool Mutate.

joakim-brannstrom commented on cmake/Modules/FindLibClang.cmake:
```diff
@@ -46,8 +46,8 @@ execute_process(COMMAND ${LLVM_CMD} version
-execute_process(COMMAND ${LLVM_CMD} cpp-flags
-    OUTPUT_VARIABLE llvm_config_CPPFLAGS
+execute_process(COMMAND ${LLVM_CMD} cxx-flags
+    OUTPUT_VARIABLE llvm_config_CXXFLAGS
```
why is it necessary to rename it from cpp to cxx?
this should only have an effect on the variable name and nothing else thus no change in the behavior

JKRT replied:
Without knowing the context I might add that CPP and CXX is usually not the same thing. Usually CPP stands for the C Pre Processor and consists of another set of flags

joakim-brannstrom commented on cmake/introspect_llvm.d:
```diff
@@ -109,11 +109,11 @@ string llvmVersion() {
 
     // -std=c++0x is required to run on travis.
-    return flags.output.strip ~ " -std=c++0x -fno-exceptions -fno-rtti";
+    return flags.output.strip ~ " -std=c++0x -finline-functions -fno-strict-aliasing -fno-exceptions -fno-rtti";
```
Should be OK to use cxxflags and probably what should be done because the c++ code that exists is mainly to interact with clang.

But remove `-std=c++0x`. Why? As the comment states it where to make it work on travis. Shouldn't be an issue anymore.
Remove `-finline-functions`. Why? Unnecessary.
Remove `-fno-strict-aliasing`. Why? If it doesn't work without this flag then it is highly likely a bug in the code. The flag is a workaround.
Remove `-fno-exceptions` and `-fno-rtti`. Let llvm-config decide if these flags (via --cxxflags) should be used or not.

joakim-brannstrom commented on libs/mutantschemata/CMakeLists.txt:
```diff
@@ -0,0 +1,58 @@
+    -I${CMAKE_SOURCE_DIR}/vendor/arsd"
+)
+
+set(deps "dextool_d2sqlite3;dextool_sumtype;dextool_miniorm;dextool_mutantschemata_cpp_source")
```
move the `dextool_mutantschemata_cpp_source` to the beginning of this dependency list. Otherwise you may end up with hard to find linking errors.

joakim-brannstrom commented on libs/mutantschemata/cpp_source/type.hpp:
```diff
@@ -0,0 +1,42 @@
+namespace CppType {
+
+struct SourceLoc {
+    unsigned line;
```
Be precise about the size of the variables by using types from `cstdint`. This is e.g. platform dependent on what size this is.

joakim-brannstrom commented on libs/mutantschemata/source/mutantschemata/api.d:
```diff
@@ -0,0 +1,151 @@
+}
+
+// D class, connection to C++ code in /cpp_source
+class SchemataApi: SchemataApiCpp {
```
I think it is enough if you add `extern(C++)` here. Should thus not need to put it on every method.

joakim-brannstrom commented on libs/mutantschemata/source/mutantschemata/externals.d:
```diff
@@ -0,0 +1,81 @@
+        void destroy();
+        void put(char);
+    }
+    extern (C++) CppBytes getBytes();
```
It is possible to write `extern(C++):`. Then everything statement after it is tagged as being `c++`.

peatersson commented:
Latest 3 commits where messy in terms of ordering, i am sorry for that. I blame the time of day.

I separated the db-handling in the mutantschemata-api. It helps both readability and helped with cleaning up the code (the comment about making all functions in the class extern).

I created a flag for the 'mutate analyze' part of Dextool Mutate so that by typing "--schemata yes", you can chose to execute mutantschemata instead of the original implementation. You don't actually need to type "yes" as a flag, you can use anything as the implementation is now.

I also cleaned up some code, removed tests in api.cpp and fixed a few other PR-comments.

joakim-brannstrom commented on libs/mutantschemata/cpp_source/api.cpp:
```diff
@@ -14,51 +14,14 @@ in order to call D code and insert/select mutants from db obtained from Dextool
+    //llvm::errs() << "Usage: rewritersample <file,otherfile,...> includeDir workingDir\n";
+    // return 1;
+    char cstr[cs.cppStr->size()+1];
+    strcpy(cstr, cs.cppStr->c_str());
```
Note that this is dangerous because cstr is stack allocated and a pointer to it escapes this function to setupClang. In this case it isn't a problem because setupClang copy the data to an internal stringstream.

`char cstr[cs.cppstr->size()+1]` will call [alloca](http://man7.org/linux/man-pages/man3/alloca.3.html) under the hood (see LLVM IR).

joakim-brannstrom commented on libs/mutantschemata/cpp_source/api.cpp:
```diff
@@ -14,51 +14,14 @@ in order to call D code and insert/select mutants from db obtained from Dextool
+    char cstr[cs.cppStr->size()+1];
+    strcpy(cstr, cs.cppStr->c_str());
 
+    setupClang(cstr, ".", ".");
```
Tip: to improve the safety of the function setupClang (make it easy to use correctly without data corruption) I recommend changing the paramters to `const std::string&`.

peatersson replied:
Sten will have to look into this since it affects rewrite.cpp (Mutant schemata implementation).

joakim-brannstrom commented on libs/mutantschemata/cpp_source/type.hpp:
```diff
@@ -17,12 +17,12 @@ C++ types for easier communication between C++ and D code.
 struct Offset {
-    unsigned begin;
-    unsigned end;
+    unsigned int begin;
```
Tip: use `uint64_t` from `cstdint` here to ensure that it is always a 64bit unsigned int.
Then on the D side use `ulong`.
Then it will work on all platforms (32 and 64 bit).

See [C and D data type compatibility](https://dlang.org/spec/interfaceToC.html#data_type_compat).

joakim-brannstrom commented on plugin/mutate/CMakeLists.txt:
```diff
@@ -5,7 +5,25 @@ list(APPEND SRC_FILES ${CMAKE_SOURCE_DIR}/plugin/source/dextool/plugin/main/stan
-set(flags "-I${CMAKE_CURRENT_LIST_DIR}/source -J${CMAKE_CURRENT_LIST_DIR}/view -I${CMAKE_SOURCE_DIR}/source -I${CMAKE_SOURCE_DIR}/dsrcgen/source -I${CMAKE_SOURCE_DIR}/clang -I${CMAKE_SOURCE_DIR}/libclang -I${CMAKE_SOURCE_DIR}/plugin/source -I${CMAKE_SOURCE_DIR}/dextool_clang_extensions/source -I${CMAKE_SOURCE_DIR}/vendor/d2sqlite3/source -J${CMAKE_SOURCE_DIR}/clang/resources -version=SqliteEnable ...
+set(flags "-I${CMAKE_SOURCE_DIR}/libs/cpptooling/source -J${CMAKE_CURRENT_LIST_DIR}/view
+    -I${CMAKE_SOURCE_DIR}/source
+    -I${CMAKE_SOURCE_DIR}/libs/dextool/source
```
aye the restructured source code location. hopefully more logical now :)

peatersson replied:
Exactly :)
I think I missed CMakeLists for schemata tho, so I pushed broken code :(

Will fix in 5

peatersson commented:
I don't know why the last azure-check failed, haven't seen that error before. I am leaving it up to Sten to fix if we want to pass all of them.

joakim-brannstrom commented:
> I don't know why the last azure-check failed, haven't seen that error before. I am leaving it up to Sten to fix if we want to pass all of them.

I think it is missing linking to `LLVMPolly.so`. I found the missing symbol in that lib by running this:
```
for F in `ls /usr/lib/llvm-8/lib/*.so`; do echo $F; readelf -Ws $F | grep -i countTrailingOnesSlowCase; done
```

joakim-brannstrom commented on libs/mutantschemata/cpp_source/rewrite.hpp:
```diff
@@ -24,23 +24,171 @@
+//#define onlyAddSub
+#define debugDuplicateMutantPrevention
+// config mutant schemata (not really implemented)
+bool negateBinaryOperators = true;
```
I would recommend moving all globals to a "context" for the schemata. This enable such things as multithreading the schemata generation, easier reset (remember the case where an application isn't possible to inject all mutants in) etc.

joakim-brannstrom commented on libs/mutantschemata/cpp_source/rewrite.hpp:
```diff
@@ -24,23 +24,171 @@
+        }
+
+        llvm::SmallString<128> indentedStr;
+        if (indentNewLines && Str.find('\n') != StringRef::npos) {
```
suggestion: Even if it looks nice to generate correctly indented code it may complicate the source code.
I would recommend reducing the "formatting" to the bare minimum. If the user need it to look pretty they can use e.g. clang-format.
(I am assuming that indentNewLines mean that the schemata do the indentation)

JKRT commented on libs/mutantschemata/cpp_source/cpp_string.cpp:
```diff
@@ -0,0 +1,51 @@
+void CppStr::put(char c){
+    cppStr->append(ONE_CHARACTER, c);
+}
+CppBytes getBytes(const char* text, int length) {
```
Length should be const here

JKRT commented on libs/mutantschemata/cpp_source/rewrite.hpp:
```diff
@@ -0,0 +1,794 @@
+
+
+// defines to control what happens \\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\\|
+// config mutant schemata
```
It seems that these booleans are used as flags. A more maintainable approach would be to assign them via macros passed via the -D flag when building dextool, then they could also be tested in separation

JKRT commented on libs/mutantschemata/cpp_source/rewrite.hpp:
```diff
@@ -0,0 +1,794 @@
+    }
+    
+    void createAndStoreActualMutant(const clang::BinaryOperator *binOp, const std::string &newExpr, const std::string &endBracket) {
+        
```
Why big empty whitespace here?

JKRT commented on libs/mutantschemata/cpp_source/rewrite.hpp:
```diff
@@ -0,0 +1,794 @@
+     */
+    bool isfileToMutate(clang::FullSourceLoc FullLocation) {
+        const clang::FileEntry *fE = SourceManager->getFileEntryForID(SourceManager->getFileID(FullLocation));
+        if (fE) {
```
The indention of these ifblocks does not follow the rest of the file where the indention is 4 spaces

JKRT commented on libs/mutantschemata/cpp_source/rewrite.hpp:
```diff
@@ -0,0 +1,794 @@
+    
+    virtual bool VisitStmt(clang::Stmt *s) {
+        clang::FullSourceLoc FullLocation = astContext->getFullLoc(s->getLocStart());
+        if (!isfileToMutate(FullLocation)) {
```
This if-statement is redundant you can just return the condition directly and take the boolean from it

JKRT replied:
Sorry. The indention confused me. However, it seems that this function never returns false? When should it do that

JKRT commented on libs/mutantschemata/cpp_source/rewrite.hpp:
```diff
@@ -0,0 +1,794 @@
+    
+    void mutateBinaryLCR(clang::BinaryOperator *binOp) {
+        switch (binOp->getOpcode()) {
+                // Logical operators
```
Instead of these comments, maybe write methods with names like handle_logical_operators and call them with suitable arguments

peatersson commented:
Latest commit was tested with a small example project. Both analyze and test-phase seems to work as intended. The only work needed was to manually add the Translation Unit constructor for the files mutated. This should however be fixed soon.

Setting environment variable is now done on the cpp-side as well, which worked just as intended.

joakim-brannstrom commented on libs/mutantschemata/cpp_source/api.cpp:
```diff
@@ -48,19 +49,8 @@ void runSchemataCpp(SchemataApiCpp *sac, CppString::CppStr cs, CppString::CppStr
-    sac->apiInsertSchemataMutant(sm);
-    */
+int setEnvironmentVariable(CppString::CppStr variable, CppString::CppStr value){
+  return setenv(variable.cppStr->c_str(), value.cppStr->c_str(), 1);
```
note: If the D side is driving the mutated binary then it is probably the call to std.process.spawnProcess that you want to inject the environment variables through.

joakim-brannstrom commented:
> Latest commit was tested with a small example project. Both analyze and test-phase seems to work as intended. The only work needed was to manually add the Translation Unit constructor for the files mutated. This should however be fixed soon.
> 
> Setting environment variable is now done on the cpp-side as well, which worked just as intended.

Aah ok. Can be improved later on. Make it work for now :)

peatersson commented:
Last commit also included some cleanup in terms of removing the use of templates (it was not needed) and unused functions.

I also added a new type for what is actually inserted into the DB, since inserting the CppStr seemed to be problematic. The solution is quite ugly, but it works for now.

peatersson commented:
Hmm, I am not sure. I just fixed it quick when i realized that the result in the DB were wrong. 

The way it is now seems reasonable, I think. When I do initialize (basically creating and setting the environment variable) I start with setting it to zero, which would mean that no mutants are executed if i run the test suite.

peatersson commented on libs/mutantschemata/source/mutantschemata/type.d:
```diff
@@ -0,0 +1,69 @@
+  Offset offset;
+  ulong status;
+  string inject;
+  AbsolutePath filePath;
```
@joakim-brannstrom 
This is where I feel you could help me with creating the checksums. The above struct is what we have in the database after analyze- and test-phase. You showed me some ideas before, but I imagine you would solve this both better and faster than me.

joakim-brannstrom replied:
let me study it over the weekend

peatersson commented on libs/mutantschemata/cpp_source/rewrite.hpp:
```diff
@@ -268,7 +275,7 @@ void writeChangedFiles(bool inPlace) {
             const clang::FileEntry *file = rewriter.getSourceMgr().getFileEntryForID(I->first);
             if (file) {
-                if (file->isValid()) {
+	      if (file->isValid()) {
```
My emacs is completely bonkers with the indentations. Certainly something we solve later with format-checkers.

joakim-brannstrom commented on libs/mutantschemata/cpp_source/rewrite.hpp:
```diff
@@ -346,11 +346,21 @@ class MutatingVisitor: public clang::RecursiveASTVisitor<MutatingVisitor> {
         sema->getDiagnostics().setErrorLimit(0);
         
+        bool rem_ok = true;
+        // TODO dirty hack to ensure that the rem operator does not cause compilation faults.
```
hehe "fulhack" ;-)

peatersson replied:
kludge ;)

peatersson commented:
Mutant Schemata is now integrated to the extent that you can run it via Dextool. This way, error handling, help when analyzing and the config-file can be used. The integration does not affect the original version of Dextool Mutate since it is "hidden" behind flags.

For now, Mutant Schemata is only implemented for the analyze/generate mutant phase and the testing phase. The supported operators are ROR, AOR and LCR.

To analyze and generate mutants with Schemata instead of original version:
`dextool mutate analyze --schemata on`
( _on_ can be changed to pretty much anything. Dextool parses the input and checks that it is not empty.)

To test the mutants with Schemata instead of original version:
`dextool mutate test --schemata on`

As an example, Schemata has been tested on Googletest in _plugin/mutate/examples/gtest_. The "manual" touch needed in order to run Schemata on it is that the _-Werror_-flag needs to be removed from the compilation (one way is to just remove it from _googletest/googletest/cmake/internal_utils.cmake_)

**NOTE:** The current implementation of Schemata rewrites your files when injecting all the mutants. Reverting back to before the insertion of mutants needs to be done manually.
